app/vmselect/promql/rollup.go

```diff
@@ -1844,8 +1844,13 @@ func rollupIncreasePure(rfa *rollupFuncArg) float64 {
 
 func rollupDelta(rfa *rollupFuncArg) float64 {
 	// There is no need in handling NaNs here, since they must be cleaned up
-	// before calling rollup funcs.
+	// before calling rollup funcs. Only StaleNaNs could remain in values - see dropStaleNaNs().
 	values := rfa.values
+	if len(values) > 0 && decimal.IsStaleNaN(values[len(values)-1]) {
+		// if last sample on interval is staleness marker then the selected series is expected
+		// to stop rendering immediately. See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8891
+		return nan
+	}
```

Comment:
interesting, this fast path won't work, as the [StaleNaN](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/8285d7d1801e1d865bdd1fb566f414c7c2ec3e04/lib/decimal/decimal.go#L398) will be changed after correction in https://github.com/VictoriaMetrics/VictoriaMetrics/blob/8285d7d1801e1d865bdd1fb566f414c7c2ec3e04/app/vmselect/promql/rollup.go#L941

See [this test](https://go.dev/play/p/d66E8rPjmmX).

Reply:
Very good point!
Looks like removeCounterResets indeed remove stale NaNs from the values. In my case, it replaces stale NaN with NaN. The fastpath doesn't trigger then, as you noticed. But calculation still remains correct - because it is enough to have NaN as last value for all `dv := vEnd - prevValue` expressions to return NaN.

Reply:
sure, just want to share since that's interesting)